### PR TITLE
War

### PR DESCRIPTION
**Pull request by Bradley-Mathi**
This is the full game code for war. Everything should be in working order in terms of code and commenting. Once this is merged to the actual branch we can complete the tests for it 👌 .

### CODE REVIEW

File: src/main/java/usecases/War.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+            this.gameOutput.sendOutput("---------------------------------------");
+
+            gameInput.stall();
+        }
+
+        if (players[0].isHandEmpty() && players[1].isHandEmpty()) {
+            this.gameOutput.sendOutput("Somehow you two have mananged to end up in an extremely improbable draw. Congratulations!");
+        } else if (players[1].isHandEmpty()) {
+            this.gameOutput.sendOutput(players[1].getName() + " is out of cards and can no longer participate. " + players[0].getName() + " wins!");
```
Comment (Luke-Avveduto) [inline src/main/java/usecases/War.java:113]: Consider pulling this out into a private method to reduce the length of this method.

File: src/main/java/usecases/War.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
+package usecases;
+
+import entities.Card;
+
+import java.util.*;
```
Comment (raylfli) [inline src/main/java/usecases/War.java:5]: Consider reducing this import to just a single class

**Review by raylfli — CHANGES_REQUESTED** [review]
Also, run the IntelliJ code formatter to fix spacing issues

File: src/main/java/usecases/War.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            this.gameOutput.sendOutput("Somehow you two have mananged to end up in an extremely improbable draw. Congratulations!");
+        } else if (players[1].isHandEmpty()) {
+            this.gameOutput.sendOutput(players[1].getName() + " is out of cards and can no longer participate. " + players[0].getName() + " wins!");
+        } else if (players[0].isHandEmpty()) {
+            this.gameOutput.sendOutput(players[0].getName() + " is out of cards and can no longer participate. " + players[1].getName() + " wins!");
+        } else {
+            this.gameOutput.sendOutput("bruh how did you get here. " + playingField[0].size() + " " + playingField[1].size());
+        }
+    }
+
```
Comment (Luke-Avveduto) [inline src/main/java/usecases/War.java:120]: Consider pulling some larger code blocks into private helper methods, such as the chunks 110-120 and 81-95

**Review by Luke-Avveduto — CHANGES_REQUESTED** [review]
Try pulling main into the branch, this should give the features you need to complete the tests. Then once the tests are added, push them and mark it for review again.

File: src/main/java/usecases/War.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+import entities.Card;
+import presenters.console.Input;
+import presenters.console.Output;
+
+import java.util.Random;
+import java.util.ArrayList;
+import java.util.List;
+
+public class War extends GameTemplate {
+    public final ArrayList[] playingField = new ArrayList[]{new ArrayList<Card>(), new ArrayList<Card>()};
```
Comment (raylfli) [inline src/main/java/usecases/War.java:12]: This should not be public. Additionally, check the IntelliJ warnings here

File: src/main/java/usecases/War.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    private final Input gameInput;
+    private final Output gameOutput;
+    private final static String[] Hierarchy = new String[]{"A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"};
```
Comment (raylfli) [inline src/main/java/usecases/War.java:15]: Variable naming convention not adhered to here; constants should be all capitals

File: src/main/java/usecases/War.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+    /**
+     * Constructor for War. Note that usernames must be a List of length 2
+     *
+     * @param gameInput  A Game.Input object allowing for player input.
+     * @param gameOutput A Game.Output object allowing for output to the player.
+     */
+    public War(Input gameInput, Output gameOutput, List<String> usernames, UserManager userManager, Random random) {
```
Comment (raylfli) [inline src/main/java/usecases/War.java:23]: Consider adding a constructor that does not have a `Random` parameter; look at `CrazyEights` for an example

**Review by raylfli — CHANGES_REQUESTED** [review]
In addition to the comments, 
* [ ] Add War to the `GameTemplate.gameFactory()` method and add War into the game list in `Main`
* [ ] Fix console output formatting
* [ ] Check all IntelliJ warnings
* [ ] Run the code formatter on all changed files

File: src/main/java/usecases/War.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+            if (winner == 2) {
+                inWar = true;
+            } else if (winner < 2) {
+                for (int j = 0; j < 2; j++) {
+                    for (int k = 0; k < playingField.get(j).size(); k++) {
+                        players[winner].addToHand(playingField.get(j).remove(k));
+                    }
+                }
+            }
```
Comment (raylfli) [inline src/main/java/usecases/War.java:95]: Weird behaviour occurring here: 
```
=======================================
              GAME SELECT              
=======================================
[1] Crazy Eights
[2] War
[0] EXIT
=======================================
Make a selection: 2
Input up to 2 usernames for players playing the game. Enter 'done' to finish.
Enter a Username: alpha
Enter a Username: beta
=======================================




=======================================
                  War
=======================================




---------------------------------------
Turn 1. Pile sizes: 0
---------------------------------------
alpha's Pile is empty
---------------------------------------
beta's Pile is empty
---------------------------------------
Press enter to continue: 
---------------------------------------
alpha flipped the card 8D
beta flipped the card AC
alpha's card outranks their opponents and they win the battle!
---------------------------------------
Press enter to continue: 
---------------------------------------
Turn 2. Pile sizes: 0
---------------------------------------
alpha's Pile is empty
---------------------------------------
beta's Pile is empty
---------------------------------------
Press enter to continue: 
---------------------------------------
alpha flipped the card 10H
beta flipped the card 9H
alpha's card outranks their opponents and they win the battle!
---------------------------------------
Press enter to continue: 
---------------------------------------
Turn 3. Pile sizes: 0
---------------------------------------
alpha's Pile is empty
---------------------------------------
beta's Pile is empty
---------------------------------------
Press enter to continue: 
---------------------------------------
alpha flipped the card 2H
beta flipped the card 2S
Both players have flipped the same rank of card, so war begins!
---------------------------------------
Press enter to continue: 
---------------------------------------
Turn 4. Pile sizes: 1
---------------------------------------
alpha's Top Card is: 2H
---------------------------------------
beta's Top Card is: 2S
---------------------------------------
Press enter to continue: 
---------------------------------------
alpha flipped the card 4C
beta flipped the card QC
beta's card outranks their opponents and they win the battle!
---------------------------------------
Press enter to continue: 
---------------------------------------
Turn 5. Pile sizes: 1
---------------------------------------
alpha's Top Card is: 2D
---------------------------------------
beta's Top Card is: 6H
---------------------------------------
Press enter to continue: 
---------------------------------------
alpha flipped the card 3C
beta flipped the card AD
alpha's card outranks their opponents and they win the battle!
---------------------------------------
Press enter to continue: ```

**Review by raylfli — CHANGES_REQUESTED** [review]
Not all cards removed off piles after a turn

**Review by raylfli — APPROVED** [review]
Concerns addressed

**Pull request merged**